---
title: "cDNA processing"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab

# Introduction
I previously processed the raw sequencing data, optimized the barcode clustering, quantified the pDNA data and normalized the cDNA data. In this script, I want to have a detailed look at the cDNA data from a general perspective.

## Description of Data
How to make a good rendering table: 
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| column1	|	column2	|	column3	|	
|----|----|----|
|1	|	2	|	3	|	
|a	|	b	|	c	|	
"
cat(tabl) # output the table in a format good for HTML/PDF/docx conversion
```

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
```

### Custom functions
Functions used thoughout this script.
```{r}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


cbind.fill <- function(...){
    nm <- list(...) 
    nm <- lapply(nm, as.matrix)
    n <- max(sapply(nm, nrow)) 
    do.call(cbind, lapply(nm, function (x) 
        rbind(x, matrix(, n-nrow(x), ncol(x))))) 
}
```


## Data import
```{r}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/results/mt20200525_reporter_activity_filt.csv", header = T)
cDNA_df$background <- as.character(cDNA_df$background)
cDNA_df <- na.omit(cDNA_df)
```

# Analysis

## First insights into data distribution - reporter activity distribution plots
```{r data_density, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## reporter activity distribution over all experiments
colors <- c("#1B998B", "#2D3047", "#FF9B71")

## reporter activity distribution for each experiment individually
ggplot(data = cDNA_df, aes(x = log_activity)) + 
  geom_density(fill = "#3D9F83") + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution") +
  xlim(-5,5) + theme_classic() + facet_wrap(vars(condition))

fill <- c("2i_neg_LIF" = "#CACDD3", 
            "2i_pos_LIF" = "#CACDD3",
            "LIF_pos_CH" = "#CACDD3",
            "LIF_pos_PD" = "#CACDD3",
            "N2B27" = "#CACDD3",
            "NPC" = "#1B998B",
            "vitA" = "#CACDD3"
            )

ggplot(data = cDNA_df, aes(x = log_activity, fill = condition, alpha = 0.1)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution") +
  scale_fill_manual(values = fill)+
  xlim(0,5) + theme_classic()

## reporter activity distribution for each experiment individually
ggplot(data = cDNA_df, aes(x = log_activity, color = condition)) + 
  stat_ecdf() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution") + ylab("fraction") +
  xlim(-5,5) + theme_classic()


## reporter activity distribution - highlight negative motif controls
ggplot(cDNA_df[cDNA_df$hPGK == "No",], aes(x = log_activity, fill = neg_ctrls, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - negative ctrls") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()


## reporter activity distribution - highlight positive hPGK control
ggplot(cDNA_df, aes(x = log_activity, fill = hPGK, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - hPGK") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()

## reporter activity distribution - highlight Miguels enhancer control
ggplot(cDNA_df, aes(x = log_activity, fill = native_enhancer, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - native enhancers") +
  scale_fill_manual(values = colors) +
  xlim(0,5) + theme_classic()+ facet_wrap(vars(condition))


## reporter activity distribution - highlight different promoters
ggplot(cDNA_df, aes(x = log_activity, fill = promoter, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - promoters") +
  scale_fill_brewer("Dark2") +
  xlim(-5,5) + theme_classic()

## reporter activity distribution - highlight random promoters
ggplot(cDNA_df, aes(x = log_activity, fill = rand_promoter, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - random promoter") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()

## reporter activity distribution - highlight different distances
ggplot(cDNA_df[cDNA_df$distance != "",], 
       aes(x = log_activity, fill = distance, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - distances") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()

## reporter activity distribution - highlight different spacings
ggplot(cDNA_df[cDNA_df$spacing != "",], aes(x = log_activity, fill = spacing, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - spacings") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()

## reporter activity distribution - highlight different backgrounds
ggplot(cDNA_df[cDNA_df$background != "0",], aes(x = log_activity, fill = background, alpha = 0.4)) + 
  geom_density() + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution - backgrounds") +
  scale_fill_manual(values = colors) +
  xlim(-5,5) + theme_classic()

## normalized bc-counts for each TF - comparing to original pDNA counts
ggplot(cDNA_df[cDNA_df$native_enhancer == "No" &
               cDNA_df$neg_ctrls  == "No" &
               cDNA_df$hPGK == "No" &
               cDNA_df$rand_promoter == "No",]) + 
  geom_density(aes(x = bc_counts_rpm, fill = "#1B998B", alpha = 0.4)) + 
  geom_density(aes(x = pDNA_counts_rpm, fill = "#2D3047", alpha = 0.4)) + 
  xlab("bc-counts per million") + 
  labs(title = "bc-count distribution cDNA (green) vs. pDNA (blue)") +
  scale_fill_manual(values = colors) +
  xlim(0,200) + theme_classic() +
  facet_wrap(~TF)

## correlation bc-counts pDNA vs. cDNA
sp <- ggscatter(cDNA_df[cDNA_df$native_enhancer == "No" &
               cDNA_df$neg_ctrls  == "No" &
               cDNA_df$hPGK == "No" &
               cDNA_df$rand_promoter == "No",],
               x = "pDNA_counts_rpm", y = "bc_counts_rpm",
               size = 0.1, alpha = 0.3, color = "#2D3047", 
   add = "reg.line",
   add.params = list(color = "blue", fill = "lightgray"), title = "normalized barcode counts cDNA vs. pDNA",
   conf.int = TRUE, ylab = "cDNA", xlab = "pDNA")
sp + stat_cor(method = "pearson", label.x = 150, label.y = 2) + 
  geom_abline(linetype = "dashed") + facet_wrap(~TF) + xlim(0,200) + ylim(0,200)


## correlation bc-counts pDNA vs. reporter activity
sp <- ggscatter(cDNA_df[cDNA_df$native_enhancer == "No" &
                    cDNA_df$neg_ctrls  == "No" &
                    cDNA_df$hPGK == "No" &
                    cDNA_df$rand_promoter == "No",],
          size = 0.1, color = "condition", alpha = 0.1,
          y = "log_activity", x = "pDNA_counts_rpm",
          title = "reporter pDNA counts vs log2 reporter activity",
          ylab = "log2 reporter activity",
          xlab = "pDNA counts per million",
          facet.by = "TF")
sp + xlim(0,200) + ylim(-5,5) + scale_color_manual(values = fill)

## let's zoom into a model TF with high noise at low pDNA: Gbx2
sp <- ggscatter(cDNA_df[cDNA_df$TF == "Gbx2" &
                    cDNA_df$neg_ctrls  == "No" &
                    cDNA_df$hPGK == "No" &
                    cDNA_df$rand_promoter == "No",],
          size = 0.1, color = "#2D3047",
          y = "log_activity", x = "pDNA_counts_rpm",
          title = "Gbx2: reporter pDNA counts vs log2 reporter activity",
          ylab = "log2 reporter activity",
          xlab = "pDNA counts per million",
          facet.by = "condition")
sp + xlim(0,200) + ylim(-5,5)

## let's zoom Stat3 because it's nice
sp <- ggscatter(cDNA_df[cDNA_df$TF == "Stat3" &
                    cDNA_df$neg_ctrls  == "No" &
                    cDNA_df$hPGK == "No" &
                    cDNA_df$rand_promoter == "No",],
          size = 0.1, color = "#2D3047",
          y = "log_activity", x = "pDNA_counts_rpm",
          title = "Stat3: reporter pDNA counts vs log2 reporter activity",
          ylab = "log2 reporter activity",
          xlab = "pDNA counts per million",
          facet.by = "condition")
sp + xlim(0,200) + ylim(-5,5)
```




## Heat map - condition vs. TF
```{r tf_activity_heatmap, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
cDNA_df$tf_activity <- ave(cDNA_df$log_reporter_activity, cDNA_df$condition, cDNA_df$TF, FUN = function(x) mean(x))

tf_activity_heatmap <- cDNA_df[cDNA_df$neg_ctrls == "No" & cDNA_df$hPGK == "No" &
                               cDNA_df$native_enhancer == "No" & cDNA_df$rand_promoter == "No",] %>% 
  select(TF, condition, tf_activity) %>% unique()
tf_activity_heatmap <- dcast(tf_activity_heatmap, TF ~ condition, value.var="tf_activity")
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="TF")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-2,2,0.04)

# pheatmap function
pheatmap(as.matrix(tf_activity_heatmap),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
```







## Heatmap per TF - comparing design activities mutated vs. non-mutated
```{r heatmap_4, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
cDNA_df$reporter <- gsub("_[0-9]{1}$", "", cDNA_df$reporter_id)
cDNA_df$tf_design_activity <- ave(cDNA_df$log_activity, 
                                  cDNA_df$condition, 
                                  cDNA_df$reporter,
                                  FUN = function(x) mean(x))


tf_activity_heatmap <- cDNA_df[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$rand_promoter == "No" &
                                 cDNA_df$hPGK == "No",] %>% 
  select(reporter, condition, tf_design_activity) %>% unique()

tf_activity_heatmap$tf_design_activity[!is.finite(tf_activity_heatmap$tf_design_activity)] <- 0
tf_activity_heatmap <- dcast(tf_activity_heatmap, reporter ~ condition,
                             value.var="tf_design_activity")
tf_activity_heatmap$TF <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_heatmap$reporter)
tf_activity_heatmap$TF <- gsub("_neg$", "", tf_activity_heatmap$TF)
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-3,3,0.06)

# pheatmap function
for(i in unique(tf_activity_heatmap$TF)){
  pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$TF == i,-8]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
}
```



## Heatmap per TF - only WT TF activities
```{r heatmap_5, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$hPGK == "No" &
                                 cDNA_df$neg_ctrls == "No",] %>% 
  select(reporter, condition, tf_design_activity) %>% unique()

tf_activity_heatmap$tf_design_activity[!is.finite(tf_activity_heatmap$tf_design_activity)] <- 0
tf_activity_heatmap <- dcast(tf_activity_heatmap, reporter ~ condition,
                             value.var="tf_design_activity")
tf_activity_heatmap$TF <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_heatmap$reporter)
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-3,3,0.06)

# pheatmap function
for(i in unique(tf_activity_heatmap$TF)){
  pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$TF == i,-8]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
}
```




## Compute activity changes relative to controls
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## TF activity relative to negative controls 
cDNA_df_ctrl <- cDNA_df[cDNA_df$native_enhancer == "No" &
                          cDNA_df$hPGK == "No" & cDNA_df$TF != "Random1" &
                          cDNA_df$TF != "Random2" & cDNA_df$TF != "Random3",] %>% 
  mutate(reporter_2 = gsub("_neg", "", reporter)) %>%
  select(reporter_2, condition, reporter_activity, neg_ctrls) %>% 
  mutate(reporter_activity = ave(reporter_activity, reporter_2, condition,
                                 neg_ctrls, FUN = function(x) mean(x))) %>%
  unique()
  

## Divide TF activity by negative control activity for each reporter
for (i in unique(cDNA_df_ctrl$reporter_2)) {
  for (j in unique(cDNA_df_ctrl$condition)) {
    if (length(cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                                     cDNA_df_ctrl$neg_ctrls == "Yes" &
                                     cDNA_df_ctrl$condition == j]) != 0 &
       length(cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                                    cDNA_df_ctrl$neg_ctrls == "No" &
                                     cDNA_df_ctrl$condition == j]) != 0) { 
        
    cDNA_df_ctrl$reporter_activity_relative[cDNA_df_ctrl$reporter_2 == i & 
                                     cDNA_df_ctrl$condition == j] <- 
      
     cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                             cDNA_df_ctrl$neg_ctrls == "No" & 
                              cDNA_df_ctrl$condition == j] /
      
      cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                              cDNA_df_ctrl$neg_ctrls == "Yes" & 
                               cDNA_df_ctrl$condition == j]
   }
    else {
     cDNA_df_ctrl$reporter_activity_relative[cDNA_df_ctrl$reporter_2 == i & 
                                       cDNA_df_ctrl$condition == j] <- 0
    }
  }
}

## Log2 transformation
cDNA_df_ctrl <- cDNA_df_ctrl %>% 
  mutate(log_reporter_activity_relative = log2(reporter_activity_relative))
  
cDNA_df_ctrl$log_reporter_activity_relative[!is.finite(cDNA_df_ctrl$log_reporter_activity_relative)] <- 0


## Generate heatmaps with relative log2 activities 
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df_ctrl[cDNA_df_ctrl$neg_ctrls == "No",] %>% 
  select(reporter_2, condition, log_reporter_activity_relative) %>% unique()

tf_activity_heatmap <- tf_activity_heatmap %>% 
  dcast(reporter_2 ~ condition, value.var = "log_reporter_activity_relative")
tf_activity_heatmap$TF <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_heatmap$reporter_2)
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter_2")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-5,5,0.1)

# pheatmap function
for(i in unique(tf_activity_heatmap$TF)){
  pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$TF == i,-8]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
}

```


















## SuperPlot of TF activity per condition
```{r SuperPlot, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Plot for each tf-activity per condition per biological replicate the techical replicate distribution and means
## Compute relevant mean and sd values
activity_mean <- cDNA_df %>% 
  select(activity, condition, rep, TF, promoter, neg_ctrls, hPGK, native_enhancer, rand_promoter) %>% 
  group_by(condition, rep, TF, promoter) %>% 
  mutate(activity = mean(activity))

activity_mean <- activity_mean %>% 
  group_by(condition, TF, promoter) %>% 
  mutate(mean = mean(activity),
         se = sd(activity)/sqrt(length(unique((activity)))))

activity_mean <- activity_mean %>% 
  mutate(upper = mean + se, lower = mean - se)


## All TFs
activity_mean_1 <- activity_mean[activity_mean$promoter == "minP" & 
                        activity_mean$neg_ctrls == "No" &
                        activity_mean$native_enhancer == "No",] %>% unique()

## hPGK controls
activity_mean_2 <- activity_mean[activity_mean$promoter == "hPGK" & 
                        activity_mean$neg_ctrls == "No" & 
                        activity_mean$native_enhancer == "No" ,] %>% unique()

## Native enhancers
activity_mean_3 <- activity_mean[activity_mean$promoter == "minP" & 
                        activity_mean$native_enhancer == "Yes",] %>% unique()


## Random promoters
activity_mean_4 <- activity_mean[activity_mean$promoter == "Random" & 
                        activity_mean$neg_ctrls == "No" & 
                        activity_mean$native_enhancer == "No",] %>% unique()

## Negative controls
activity_mean_5 <- activity_mean[activity_mean$promoter == "minP" & 
                        activity_mean$neg_ctrls == "Yes" & 
                        activity_mean$native_enhancer == "No",] %>% unique()


## Generate SuperPlot
### Selection 1 - All TFs
ggplot(cDNA_df[cDNA_df$promoter == "minP" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$native_enhancer == "No",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean_1, 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean_1, aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean_1, 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition", 
         subtitle = "Total data distribution") +
    facet_wrap(~TF) +
    theme_bw() +
    ylim(0,6) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))



### Selection 2 - hPGK controls
ggplot(cDNA_df[cDNA_df$promoter == "hPGK" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$native_enhancer == "No",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean_2, 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean_2, aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean_2, 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition", 
         subtitle = "Total data distribution") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))



### Selection 3 - Native enhancers
ggplot(cDNA_df[cDNA_df$promoter == "minP" & 
                        cDNA_df$native_enhancer == "Yes",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean_3, 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean_3, aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean_3, 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition", 
         subtitle = "Total data distribution") +
    facet_wrap(~TF) +
    ylim(0,6) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))




### Selection 4 - Random promoter
ggplot(cDNA_df[cDNA_df$promoter == "Random" & 
                        cDNA_df$neg_ctrls == "No" & 
                        cDNA_df$native_enhancer == "No",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean_4, 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean_4, aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean_4, 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition - Random promoter", 
         subtitle = "Total data distribution") +
    facet_wrap(~TF) +
    theme_bw() +
    ylim(0,6) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))



### Selection 5 - All TF-negative controls
ggplot(cDNA_df[cDNA_df$promoter == "minP" & 
                        cDNA_df$neg_ctrls == "Yes" &
                        cDNA_df$native_enhancer == "No",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean_5, 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean_5, aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean_5, 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition - only negative controls", 
         subtitle = "Total data distribution") +
    facet_wrap(~TF) +
    theme_bw() +
    ylim(0,6) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))
```


# SuperPlots comparing different designs 
```{r superplot_2, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Same as before - now compare different designs
## Generate facet_wrap plot for each TF - each facet is a different design 
## Compute values for single and combinatorial designs


## Compute relevant mean and sd values
activity_mean <- cDNA_df %>% 
  select(activity, condition, rep, TF, promoter, neg_ctrls, 
         hPGK, native_enhancer, rand_promoter, reporter_id, spacing, 
         distance, background) %>% 
  group_by(condition, rep, reporter_id) %>% 
  mutate(activity = mean(activity))

activity_mean <- activity_mean %>% 
  group_by(condition, reporter_id) %>% 
  mutate(mean = mean(activity),
         se = sd(activity)/sqrt(length(unique((activity)))))

activity_mean <- activity_mean %>% 
  mutate(upper = mean + se, lower = mean - se)

activity_mean <- activity_mean[activity_mean$neg_ctrls == "No" &
                        activity_mean$native_enhancer == "No",] %>% unique()


## Generate SuperPlot
### Selection 1 - All Designs separately
ggplot(cDNA_df[cDNA_df$TF == "Trp53" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$native_enhancer == "No",], 
              aes(x=condition, y=activity, color=factor(rep))) + 
    scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
    geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
    geom_errorbar(data = activity_mean[activity_mean$TF == "Trp53",], 
                  aes(ymin = lower, 
                      ymax = upper), colour = "black") + 
    geom_quasirandom(data = activity_mean[activity_mean$TF == "Trp53",], aes(alpha = 1), size=4) + 
    stat_compare_means(data=activity_mean[activity_mean$TF == "Trp53",], 
                       comparisons = list(c(1:2)), 
                   method="t.test", paired=TRUE) + 
    labs(title = "TF activity distribution per condition", 
         subtitle = "Total data distribution") +
    facet_wrap(~reporter_id) +
    theme_bw() +
    ylim(0,16) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
    ylab("TF reporter activity") + xlab("") + 
    theme(text = element_text(size = 12))


## Subdivide in different designs 
### Spacing and distance
cDNA_df2 <- cDNA_df %>% 
  mutate(space_dist = paste(spacing, distance)) %>% 
  select(activity, condition, TF, space_dist, neg_ctrls, native_enhancer, promoter) 

cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
                       cDNA_df2$native_enhancer == "No" & 
                       cDNA_df2$promoter != "Random" &
                       cDNA_df2$promoter != "hPGK",] %>% unique()



for(i in 1:4){
  p <- ggplot(cDNA_df2, 
        aes(x=condition, y=activity, color=space_dist)) + 
      scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
      geom_quasirandom(dodge.width = 0.9) +
      geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
      labs(title = "TF activity distribution per condition", 
           subtitle = "Total data distribution") +
      facet_wrap_paginate(~TF, nrow = 3, ncol = 3, page = i) +
      theme_bw() +
      ylim(0,6) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
           axis.text.y = element_text(size = 10)) +
      ylab("TF reporter activity") + xlab("") + 
      theme(text = element_text(size = 12))
  print(p)
}



### Background
cDNA_df2 <- cDNA_df %>% 
  select(activity, condition, TF, background, neg_ctrls, native_enhancer, promoter) 


cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
                       cDNA_df2$native_enhancer == "No" & 
                       cDNA_df2$promoter != "Random" &
                       cDNA_df2$promoter != "hPGK",] %>% unique()



for(i in 1:4){
  p <- ggplot(cDNA_df2, 
        aes(x=condition, y=activity, color=background)) + 
      scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
      geom_quasirandom(dodge.width = 0.9) +
      geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
      labs(title = "TF activity distribution per condition", 
           subtitle = "Total data distribution") +
      facet_wrap_paginate(~TF, nrow = 3, ncol = 3, page = i) +
      theme_bw() +
      ylim(0,6) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
           axis.text.y = element_text(size = 10)) +
      ylab("TF reporter activity") + xlab("") + 
      theme(text = element_text(size = 12))
  print(p)
}


### Promoter
cDNA_df2 <- cDNA_df %>% 
  select(activity, condition, TF, promoter, neg_ctrls, native_enhancer, promoter) 


cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
                       cDNA_df2$native_enhancer == "No" &
                       cDNA_df2$promoter != "hPGK",] %>% unique()



for(i in 1:4){
  p <- ggplot(cDNA_df2, 
        aes(x=condition, y=activity, color=promoter)) + 
      scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
      geom_quasirandom(dodge.width = 0.9) +
      geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
      labs(title = "TF activity distribution per condition", 
           subtitle = "Total data distribution") +
      facet_wrap_paginate(~TF, nrow = 3, ncol = 3, page = i) +
      theme_bw() +
      ylim(0,6) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
           axis.text.y = element_text(size = 10)) +
      ylab("TF reporter activity") + xlab("") + 
      theme(text = element_text(size = 12))
  print(p)
}


### Promoter - native enhancers
cDNA_df2 <- cDNA_df %>% 
  select(activity, condition, TF, background, neg_ctrls, native_enhancer, promoter) 


cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
                       cDNA_df2$native_enhancer == "Yes" &
                       cDNA_df2$promoter != "hPGK",] %>% unique()



for(i in 1:4){
  p <- ggplot(cDNA_df2, 
        aes(x=condition, y=activity, color=promoter)) + 
      scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
      geom_quasirandom(dodge.width = 0.9) +
      geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
      labs(title = "TF activity distribution per condition", 
           subtitle = "Total data distribution") +
      facet_wrap_paginate(~TF, nrow = 3, ncol = 3, page = i) +
      theme_bw() +
      ylim(0,6) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
           axis.text.y = element_text(size = 10)) +
      ylab("TF reporter activity") + xlab("") + 
      theme(text = element_text(size = 12))
  print(p)
}
```








## Log-linear expression modelling to explain variance - model for each TF - only WT
```{r log-linear-model-3, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
### For each TF separately - only the non-mutated 
cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, condition, FUN = function(x) mean(x))) %>%
  select(log_reporter_activity, TF, condition, promoter, spacing, distance, background) %>%
  unique()

### Calculate expression variance
exp_variance <- data.frame(c("condition", "promoter", 
                             "spacing", "distance", "background", "Residuals"))
prediction <- data.frame(1:504)
names(prediction) <- "id"
names(exp_variance) <- "feature"
weight <- data.frame(c("(Intercept)", "condition2i_pos_LIF", 
                  "conditionLIF_pos_CH", "conditionLIF_pos_PD", "conditionN2B27", 
                  "conditionNPC", "conditionvitA", "promotermCMV", "promoterminP", 
                  "promoterRandom",
                  "spacing10bp", "distance21bp", "background2", "background3"))
names(weight) <- "feature"


cDNA_df_TF$row <- rownames(cDNA_df_TF)



for (i in unique(cDNA_df_TF$TF)) {
  x <- lm(log_reporter_activity ~ condition + 
                       promoter + spacing + distance + background, 
                     cDNA_df_TF[cDNA_df_TF$TF == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  par(mfrow=c(2,2))
  plot(x, main = i)
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Fine-tune dfs
exp_variance <- melt(exp_variance, variable.name = "TF")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% select(-id) %>% melt(variable.name = "TF") %>% na.omit() %>%
  select(-TF) %>% setnames("value", "activity_predicted")
cDNA_df_TF <- merge(cDNA_df_TF, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "TF")
categorie <- c("mutated", "condition", "promoter", 
               "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- weight$features
weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight <- weight %>% select(-feature)


# Model parameters
ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)

# Model prediction accuracy
sp <- ggscatter(cDNA_df_TF, x = "log_reporter_activity", y = "activity_predicted",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "Predicted log2-activity", xlab = "Average log2-activity",
   title = "Correlation between predicted and actual reporter activities") +
  facet_wrap(~TF) + theme_bw()
sp + stat_cor(method = "pearson", label.x = -3, label.y = 3.7) + geom_abline(linetype = "dashed")


## Plot weights of parameters in model - this helps to understand the model 
level_order <- c("2i_pos_LIF", "LIF_pos_CH", 
                 "LIF_pos_PD", "vitA", "N2B27", 
                 "NPC", "mCMV", "minP", "Random",
                 "2", "3", "10bp", "21bp")


ggplot(weight, 
       aes(x = factor(feature, level = level_order), y = value, fill = condition)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)



## Final overview - model fit for each TF
# Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature)

# Model parameters
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total variance explained"))

# Include expression data
cDNA_df_sel <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No",] %>% 
  select(TF, condition, promoter, spacing, 
                                  distance, background, reporter_activity) %>% unique()
activity_condition <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, condition, FUN = function(x) mean(x))) %>%
  select(TF, condition, activity) %>%
  setnames("condition", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "condition") %>% 
  select(-sum)
activity_promoter <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, promoter, FUN = function(x) mean(x))) %>%
  select(TF, promoter, activity) %>%
  setnames("promoter", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "promoter")%>% 
  select(-sum)
activity_spacing <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, spacing, FUN = function(x) mean(x))) %>%
  select(TF, spacing, activity) %>%
  setnames("spacing", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "spacing")%>% 
  select(-sum)
activity_distance <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, distance, FUN = function(x) mean(x))) %>%
  select(TF, distance, activity) %>%
  setnames("distance", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "distance")%>% 
  select(-sum)
activity_background <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, background, FUN = function(x) mean(x))) %>%
  select(TF, background, activity) %>%
  setnames("background", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "background")%>% 
  select(-sum)


exp_variance <- exp_variance %>% 
  select(-group) %>%
  dcast(TF ~ feature, value.var = "value") %>%
  setnames(c("background", "condition", "distance", "promoter", "spacing"), c("background_total", "condition_total", "distance_total", "promoter_total", "spacing_total"))

activity_condition <- merge(activity_condition, exp_variance %>% select(TF, condition_total))
activity_condition$activity <- activity_condition$activity * activity_condition$condition_total
activity_condition <- activity_condition %>% select(-condition_total)

activity_promoter <- merge(activity_promoter, exp_variance %>% select(TF, promoter_total))
activity_promoter$activity <- activity_promoter$activity * activity_promoter$promoter_total
activity_promoter <- activity_promoter %>% select(-promoter_total)

activity_spacing <- merge(activity_spacing, exp_variance %>% select(TF, spacing_total))
activity_spacing$activity <- activity_spacing$activity * activity_spacing$spacing_total
activity_spacing <- activity_spacing %>% select(-spacing_total)


activity_distance <- merge(activity_distance, exp_variance %>% select(TF, distance_total))
activity_distance$activity <- activity_distance$activity * activity_distance$distance_total
activity_distance <- activity_distance %>% select(-distance_total)

activity_background <- merge(activity_background, exp_variance %>% select(TF, background_total))
activity_background$activity <- activity_background$activity * activity_background$background_total
activity_background <- activity_background %>% select(-background_total)

total_variance <- exp_variance %>% select(TF, `total variance explained`) %>%
  mutate(condition = "all",
         group = "total") %>%
  setnames("total variance explained", "activity")


x <- rbind(activity_condition,
           activity_promoter, activity_spacing,
           activity_distance, activity_background, total_variance)


colors <- c("2i_neg_LIF" = "#2D3047", 
            "2i_pos_LIF" = "#56596C",
            "LIF_pos_CH" = "#808291",
            "LIF_pos_PD" = "#A9ABB5",
            "N2B27" = "#BABCC4",
            "NPC" = "#CACDD3",
            "vitA" = "#DBDDE1",
            
            "hBGm" = "#2D3047",
            "mCMV" = "#56596C",
            "minP" = "#808291",
            "Random" = "#A9ABB5",
            
            "10bp" = "#2D3047",
            "5bp" = "#56596C",
            
            "21bp" = "#56596C",
            
            "2" = "#2D3047",
            "1" = "#56596C",
            "3" = "#808291",
            
            "all" = "#1B998B"
            )

ggplot(x, aes(fill=condition, y=activity, x=reorder(group, activity))) +
  geom_bar(position="stack", stat="identity") +
  coord_flip() +
  scale_fill_manual(values = colors) +
  #scale_fill_viridis(discrete = T) +
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)
```




## Log-linear expression modelling to explain variance - model for each TF - only WT - without condition
```{r log-linear-model-4, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
### For each TF separately - only the non-mutated 
cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$rand_promoter == "No" &
                        cDNA_df$neg_ctrls == "No",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, FUN = function(x) mean(x))) %>%
  select(log_reporter_activity, TF, promoter, spacing, distance, background) %>%
  unique()

### Calculate expression variance
exp_variance <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals"))
prediction <- data.frame(1:504)
names(prediction) <- "id"
names(exp_variance) <- "feature"
weight <- data.frame(c("(Intercept)", "promotermCMV", "promoterminP", 
                  "spacing5bp", "distance21bp", "background2", "background3"))
names(weight) <- "feature"


cDNA_df_TF$row <- rownames(cDNA_df_TF)

for (i in unique(cDNA_df_TF$TF)) {
  x <- lm(log_reporter_activity ~ promoter + spacing + distance + background, 
                     cDNA_df_TF[cDNA_df_TF$TF == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Fine-tune dfs
exp_variance <- melt(exp_variance, variable.name = "TF")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% select(-id) %>% melt(variable.name = "TF") %>% na.omit() %>%
  select(-TF) %>% setnames("value", "activity_predicted")
cDNA_df_TF <- merge(cDNA_df_TF, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "TF")
categorie <- c("mutated", "condition", "promoter", 
               "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- weight$features
weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight <- weight %>% select(-feature)


# Model parameters
ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)

# Model prediction accuracy
sp <- ggscatter(cDNA_df_TF, x = "log_reporter_activity", y = "activity_predicted",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "Predicted log2-activity", xlab = "Average log2-activity",
   title = "Correlation between predicted and actual reporter activities") +
  facet_wrap(~TF) + theme_bw()
sp + stat_cor(method = "pearson", label.x = -3, label.y = 3.7) + geom_abline(linetype = "dashed")


## Plot weights of parameters in model - this helps to understand the model 
level_order <- c("mCMV", "minP", 
                 "2", "3", "5bp", "21bp")


ggplot(weight, 
       aes(x = factor(feature, level = level_order), y = value, fill = condition)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)



## Final overview - model fit for each TF
# Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature)

# Model parameters
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total variance explained"))

# Include expression data
cDNA_df_sel <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No",] %>% 
  select(TF,  promoter, spacing, 
                                  distance, background, reporter_activity) %>% unique()

activity_promoter <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, promoter, FUN = function(x) mean(x))) %>%
  select(TF, promoter, activity) %>%
  setnames("promoter", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "promoter")%>% 
  select(-sum)
activity_spacing <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, spacing, FUN = function(x) mean(x))) %>%
  select(TF, spacing, activity) %>%
  setnames("spacing", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "spacing")%>% 
  select(-sum)
activity_distance <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, distance, FUN = function(x) mean(x))) %>%
  select(TF, distance, activity) %>%
  setnames("distance", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "distance")%>% 
  select(-sum)
activity_background <- cDNA_df_sel %>% 
  mutate(activity = ave(reporter_activity, TF, background, FUN = function(x) mean(x))) %>%
  select(TF, background, activity) %>%
  setnames("background", "condition") %>%
  unique() %>% 
  mutate(sum = ave(activity, TF, FUN = function(x) sum(x)),
         activity = activity / sum,
         group = "background")%>% 
  select(-sum)


exp_variance <- exp_variance %>% 
  select(-group) %>%
  dcast(TF ~ feature, value.var = "value") %>%
  setnames(c("background", "distance", "promoter", "spacing"), c("background_total", "distance_total", "promoter_total", "spacing_total"))

activity_promoter <- merge(activity_promoter, exp_variance %>% select(TF, promoter_total))
activity_promoter$activity <- activity_promoter$activity * activity_promoter$promoter_total
activity_promoter <- activity_promoter %>% select(-promoter_total)

activity_spacing <- merge(activity_spacing, exp_variance %>% select(TF, spacing_total))
activity_spacing$activity <- activity_spacing$activity * activity_spacing$spacing_total
activity_spacing <- activity_spacing %>% select(-spacing_total)


activity_distance <- merge(activity_distance, exp_variance %>% select(TF, distance_total))
activity_distance$activity <- activity_distance$activity * activity_distance$distance_total
activity_distance <- activity_distance %>% select(-distance_total)

activity_background <- merge(activity_background, exp_variance %>% select(TF, background_total))
activity_background$activity <- activity_background$activity * activity_background$background_total
activity_background <- activity_background %>% select(-background_total)

total_variance <- exp_variance %>% select(TF, `total variance explained`) %>%
  mutate(condition = "all",
         group = "total") %>%
  setnames("total variance explained", "activity")


x <- rbind(activity_promoter, activity_spacing,
           activity_distance, activity_background, total_variance)


colors <- c("hBGm" = "#2D3047",
            "mCMV" = "#56596C",
            "minP" = "#808291",
            "Random" = "#A9ABB5",
            
            "10bp" = "#2D3047",
            "5bp" = "#56596C",
            
            "21bp" = "#56596C",
            
            "2" = "#2D3047",
            "1" = "#56596C",
            "3" = "#808291",
            
            "all" = "#1B998B"
            )

ggplot(x, aes(fill=condition, y=activity, x=reorder(group, activity))) +
  geom_bar(position="stack", stat="identity") +
  coord_flip() +
  scale_fill_manual(values = colors) +
  #scale_fill_viridis(discrete = T) +
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)
```









## Correlation between 2i TF activity and 2i expression levels
```{r log-linear-model-5, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Incorporate RNA-seq data of mESCs in 2i+LIF
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

# remove Serum data and calculate mean expression
tib_fpkm_joshi$twoi <- (tib_fpkm_joshi$twoi_1 + tib_fpkm_joshi$twoi_2)/2
tib_fpkm_joshi <- tib_fpkm_joshi %>% select(symbol, twoi) %>%
  filter(symbol %in% c("NFE2L2", "KLF4", "NFYA", "SP1", "SMAD4", "ZFX", "IRF3", "CREB1", "ELK1", 
                    "ZFP42", "OTX1", "GLI1", "NFKB1", "GBX2", "TFCP2L1", "TCF7", "RARA", 
                    "STAT3", "RBPJ", "TRP53", "ESRRB", "NEUROG2", "PAX6", "NR3C1", "IRX3",
                    "HOMEZ"))

names <- data.frame(c("NFE2L2", "KLF4", "NFYA", "SP1", "SMAD4", "ZFX", "IRF3", "CREB1", "ELK1", 
                    "ZFP42", "OTX1", "GLI1", "NFKB1", "GBX2", "TFCP2L1", "TCF7", "RARA", 
                    "STAT3", "RBPJ", "TRP53", "ESRRB", "NEUROG2", "PAX6", "NR3C1", "IRX3",
                    "HOMEZ"),
           c("Nfe2l2", "Klf4", "Nfya", "Sp1", "Smad4", "Zfx", "Irf3", "Creb1", "Elk1", 
                    "Zfp42", "Otx1", "Gli1", "Nfkb1", "Gbx2", "Tcfcp2l1", "Tcf7l2", "Rara", 
                    "Stat3", "Rbpj", "Trp53", "Esrrb", "Neurog2", "Pax6", "Nr3c1(GR)", "Irx3",
                    "Homez"))
names(names) <- c("symbol", "TF")
tib_fpkm_joshi <- merge(tib_fpkm_joshi, names) %>%
  select(-symbol)
cDNA_df <- merge(cDNA_df, tib_fpkm_joshi, all = T)


### For each TF separately - only the non-mutated 
cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$rand_promoter == "No" &
                        cDNA_df$neg_ctrls == "No" & cDNA_df$condition == "2i_pos_LIF",] %>% 
  select(tf_activity, TF, twoi) %>%
  unique() %>% na.omit()


# Model prediction accuracy
sp <- ggscatter(cDNA_df_TF, x = "tf_activity", y = "twoi",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "TF expression in 2i (FPKM)", xlab = "Measured TF activity in 2i",
   title = "Correlation between predicted and actual reporter activities") + theme_bw()
sp + stat_cor(method = "pearson", label.x = -3, label.y = 3.7) + geom_abline(linetype = "dashed")

```



## Exporting potential data. 
```{r}
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

